File: cci05.01-insert-into-bits-lcci/cpp/Solution.cpp
class Solution {
public:
    int insertBits(int N, int M, int i, int j) {
        vector<int> nbits(32);
        vector<int> mbits(32);
        int nbitsNum = 0;
        int mbitsNum = 0;
        long mask = 1;
        for (int k = 0; k < 32; ++k) {
            if ((N & mask) != 0) {
                nbits[k] = 1;
            }
            mask <<= 1;
        }
        mask = 1L;
        for (int k = 0; k < 32; ++k) {
            if ((M & mask) != 0) {
                mbits[k] = 1;
            }
            mask <<= 1;
        }
        for (int k = i; k <= j; ++k) {
            nbits[k] = mbits[k - i];
        }
        mask = 1;
        int ret = 0;
        for (int k = 0; k < 32; ++k) {
            ret += (nbits[k] * mask);
            mask <<= 1;
        }
        return ret;
    }
};

File: cci05.01-insert-into-bits-lcci/python/Solution.py
class Solution:
    def insertBits(self, N: int, M: int, i: int, j: int) -> int:
        nbits = [0] * 32
        mbits = [0] * 32
        nbitsNum = 0
        mbitsNum = 0
        mask = 1
        for k in range(32):
            if (N & mask) != 0:
                nbits[k] = 1
            mask <<=1
        mask = 1
        for k in range(32):
            if (M&mask) != 0:
                mbits[k] = 1
            mask <<= 1
        for k in range(i,j+1):
            nbits[k] = mbits[k-i]
        mask = 1
        ret = 0
        for k in range(32):
            ret += (nbits[k] * mask)
            mask <<= 1
        return ret
